substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/parallel/ParallelGCImpl.java

```diff
@@ -0,0 +1,258 @@
+                    try {
+                        do {
+                            debugLog().string("WW block ").unsigned(n).newline();
+                            mutex.lock();
+                            parPhase.block();
+                            mutex.unlock();
+                        } while ((busyWorkers.get() & (1 << n)) == 0);
+
+                        debugLog().string("WW run ").unsigned(n).newline();
+                        drainBuffer();
+                        int witness = busyWorkers.get();
+                        int expected;
+                        do {
+                            expected = witness;
+                            witness = busyWorkers.compareAndExchange(expected, expected & ~(1 << n));
+                        } while (witness != expected);
+                        if (busyWorkers.get() == 0) {
+                            mutex.lock();
+                            seqPhase.signal();
+                            mutex.unlock();
+                        }
+                        debugLog().string("WW idle ").unsigned(n).newline();
+                    } catch (Throwable ex) {
+                        VMError.shouldNotReachHere(ex);
+                    }
+                }
+        });
+        t.setName("ParallelGCWorker-" + n);
+        t.setDaemon(true);
+        t.start();
+    }
+
+    private void waitForIdleImpl() {
+        inParallelPhase = true;
+
+        debugLog().string("PP start workers\n");
+        busyWorkers.set(allWorkersBusy);
+        parPhase.broadcast();     // let worker threads run
+
+        drainBuffer();
+
+        mutex.lock();
+        while (busyWorkers.get() != 0) {
+            debugLog().string("PP wait busy=").unsigned(busyWorkers.get()).newline();
+            seqPhase.block();     // wait for them to become idle
+        }
+        mutex.unlock();
+
+        inParallelPhase = false;
+    }
+
+    private void drainBuffer() {
+        while (true) {
+            Pointer ptr;
+            while ((ptr = buffer.pop()).notEqual(WordFactory.nullPointer())) {
+                debugLog().string("WW drain chunk=").zhex(ptr).newline();
+                if (ptr.and(UNALIGNED_BIT).notEqual(0)) {
+                    UnalignedHeapChunk.walkObjectsInline((UnalignedHeapChunk.UnalignedHeader) ptr.and(~UNALIGNED_BIT), getVisitor());
+                } else {
+                    AlignedHeapChunk.walkObjectsInline((AlignedHeapChunk.AlignedHeader) ptr, getVisitor());
+                }
+            }
+            AlignedHeapChunk.AlignedHeader tlab = allocChunkTL.get();
+            debugLog().string("WW drain tlab=").zhex(tlab).newline();
+            if (tlab.equal(WordFactory.nullPointer())) {
+                break;
+            } else {
+                scannedChunkTL.set(tlab);
+                AlignedHeapChunk.walkObjectsInline(tlab, getVisitor());
+                if (allocChunkTL.get().equal(tlab)) {
+                    // this tlab is now black, retire it
```

Comment:
It does: in my testing, I sometimes get 40 or so retired chunks using 8 parallel threads. One chunk per thread is necessarily left partially filled at the end of GC when there are no more objects to copy, so wasted are 32 chunks == about 16 Mb. I need to rethink the algorithm, probably store a position of the last scanned object in the `allocChunkTL` so it can be filled up to the end and rescanned afterwards. Another idea is to put individual objects rather than chunks in the buffer, that would eliminate this problem completely.

Reply:
I've implemented the first idea -- offset of the last black object in a chunk is stored in a thread local variable, any subsequent scans of the chunk start from that offset. It complicates things a bit, but not too much. Max RSS has decreased as a result and is now about the same as for serial GC.